src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1663,27 +1664,34 @@ public SolutionState WithFrozenPartialCompilationIncludingSpecificDocument(Docum
-                        || _documentIdOfLatestSolutionWithPartialCompilation != documentId)
+                    var reuseExistingPartialSolution =
+                        currentPartialSolution != null &&
+                        (DateTime.UtcNow - _timeOfLatestSolutionWithPartialCompilation).TotalSeconds < 0.1 &&
+                        _documentIdOfLatestSolutionWithPartialCompilation == documentId;
+
+                    if (reuseExistingPartialSolution)
                     {
-                        var tracker = this.GetCompilationTracker(documentId.ProjectId);
-                        var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+                        SolutionLogger.UseExistingPartialSolution();
+                        return currentPartialSolution;
+                    }
 
-                        var newIdToProjectStateMap = _projectIdToProjectStateMap.SetItem(documentId.ProjectId, newTracker.ProjectState);
-                        var newIdToTrackerMap = _projectIdToTrackerMap.SetItem(documentId.ProjectId, newTracker);
+                    // if we don't have one or it is stale, create a new partial solution
+                    var tracker = this.GetCompilationTracker(documentId.ProjectId);
+                    var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
 
-                        currentPartialSolution = this.Branch(
-                            idToProjectStateMap: newIdToProjectStateMap,
-                            projectIdToTrackerMap: newIdToTrackerMap,
-                            dependencyGraph: CreateDependencyGraph(_projectIds, newIdToProjectStateMap));
+                    var newIdToProjectStateMap = _projectIdToProjectStateMap.SetItem(documentId.ProjectId, newTracker.ProjectState);
+                    var newIdToTrackerMap = _projectIdToTrackerMap.SetItem(documentId.ProjectId, newTracker);
 
-                        _latestSolutionWithPartialCompilation = new WeakReference<SolutionState>(currentPartialSolution);
-                        _timeOfLatestSolutionWithPartialCompilation = DateTime.UtcNow;
-                        _documentIdOfLatestSolutionWithPartialCompilation = documentId;
-                    }
+                    currentPartialSolution = this.Branch(
+                        idToProjectStateMap: newIdToProjectStateMap,
+                        projectIdToTrackerMap: newIdToTrackerMap,
+                        dependencyGraph: CreateDependencyGraph(_projectIds, newIdToProjectStateMap));
+
+                    _latestSolutionWithPartialCompilation = new WeakReference<SolutionState>(currentPartialSolution);
+                    _timeOfLatestSolutionWithPartialCompilation = DateTime.UtcNow;
+                    _documentIdOfLatestSolutionWithPartialCompilation = documentId;
 
+                    SolutionLogger.CreatePartialSolution();
                     return currentPartialSolution;
```

Comment:
comment https://github.com/dotnet/roslyn/pull/31478/files?w=1#r238007632 also applies here.  Current approach is brittle as it is trivial for people to change flow or add new exitpoints without them (or reviewers) realizing there is a telemetry requirement as well.  Caller/callee separation enforces the requirement that all return-flow points need to be cognizant of the telemetry requirement.